### PR TITLE
UCX plugin

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
This PR adds support for running NCCL over UCX. In order to do this
1) IB plugin and SHARP plugin separated, SHARP plugin can use OOB based on either verbs or ucx implementation.
2) Added configure parameter to build with or without ucx.
3) Plugin init function moved to utils and during first call it chooses what underlying layer verbs or ucx to use. User can control this behavior with NCCL_PLUGIN_P2P environment variable. Verbs plugin is default.
4) Added tests for ucx plugin and combination of sharp + ucx for out of band.

### CODE REVIEW

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI status: <b>UNKNOWN</b> (click for details)</summary>

**Note:** the artefacts will be deleted after 28-Jan-2020

Artefact file | Status
-- | --
[build_nccl_rdma_sharp_plugins.log](https://gist.githubusercontent.com/mellanox-github/5954708d1cdd6d2453f6d58f1b6fd770/raw/d89c2ff8a1547f67367099539744718b9c5e927a/build_nccl_rdma_sharp_plugins.log) | :heavy_check_mark: PASS
[configure_sharp_startup.log](https://gist.githubusercontent.com/mellanox-github/5954708d1cdd6d2453f6d58f1b6fd770/raw/6e9b315a0411116096258cb5ed80b7c0c64f3b3a/configure_sharp_startup.log) | :heavy_check_mark: PASS
[run_nccl_test.log](https://gist.githubusercontent.com/mellanox-github/5954708d1cdd6d2453f6d58f1b6fd770/raw/6eb6b280abbbeb140b8c2d7b86af5ccb128562cc/run_nccl_test.log) | :question: UNKNOWN

**Comment by mellanox-github** [comment]
Mellanox CI FAILed.

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI status: <b>PASS</b#  > (click for details)</summary>

**Note:** the artefacts will be deleted after 31-Jan-2020

Artefact file | Status
-- | --
[build_nccl_rdma_sharp_plugins.log](https://gist.githubusercontent.com/mellanox-github/667827996b89275933ea370c1700ddb2/raw/3ea47cd293e08e8b8f3d40820b95e7d78693fba8/build_nccl_rdma_sharp_plugins.log) | :heavy_check_mark: PASS
[configure_sharp_startup.log](https://gist.githubusercontent.com/mellanox-github/667827996b89275933ea370c1700ddb2/raw/b39768babef9fa999774177965f1c443c92d6f4f/configure_sharp_startup.log) | :heavy_check_mark: PASS
[configure_sharp_stop.log](https://gist.githubusercontent.com/mellanox-github/667827996b89275933ea370c1700ddb2/raw/00820a425bde73b4dac4f278f8e3dc538229fe6a/configure_sharp_stop.log) | :heavy_check_mark: PASS
[run_nccl_test.log](https://gist.githubusercontent.com/mellanox-github/667827996b89275933ea370c1700ddb2/raw/8fcaa797af4b4fd37c98876d3191e6ee373f9cc7/run_nccl_test.log) | :heavy_check_mark: PASS

**Comment by mellanox-github** [comment]
Mellanox CI PASSed.

File: src/ucx_plugin.c [hunk 1]
```diff
@@ -0,0 +1,769 @@
...
+      wait_close(comm->worker, close_req);  
+    }
+    if (comm->gpuFlush.enabled) {
+      close_req = ucp_ep_close_nb(comm->gpuFlush.flush_ep, UCP_EP_CLOSE_MODE_FLUSH);
+      wait_close(comm->worker, close_req);
+    }
+
+    close(comm->fd);
+    free(comm);
+  }
```
Comment (bureddy) [inline src/ucx_plugin.c]: comm->worker_refcount--;
if (comm->worker->refcount == 0) {
   oob_allgather();
   ucp_worker_destroy();
   ucp_finalize();
}

File: src/ucx_plugin.c [hunk 2]
```diff
@@ -0,0 +1,769 @@
...
+      ucp_worker_progress(comm->worker);
+    }
+    req->completed = 0;
+    ucp_request_release(req);
+  }
+  ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS; //|
+  //                         UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+  ep_params.address = (ucp_address_t*)(msg + 1);
+  //  ep_params.err_mode        = err_handling_opt.ucp_err_mode;
+  UCXCHECK(ucp_ep_create(comm->worker, &ep_params, &comm->ep));
```
Comment (bureddy) [inline src/ucx_plugin.c]: comm->worker->refcount++;

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI status: <b>PASS</b#  > (click for details)</summary>

**Note:** the artefacts will be deleted after 05-Feb-2020

Artefact file | Status
-- | --
[build_nccl_rdma_sharp_plugins.log](https://gist.githubusercontent.com/mellanox-github/f88dbb2daac122e356460f7e9edc8f85/raw/5366608c043f452ffb66a29a4787a4664306702a/build_nccl_rdma_sharp_plugins.log) | :heavy_check_mark: PASS
[configure_sharp_startup.log](https://gist.githubusercontent.com/mellanox-github/f88dbb2daac122e356460f7e9edc8f85/raw/025ec4000e5bbf374221bc929a7da85e272eadd1/configure_sharp_startup.log) | :heavy_check_mark: PASS
[configure_sharp_stop.log](https://gist.githubusercontent.com/mellanox-github/f88dbb2daac122e356460f7e9edc8f85/raw/b79b72541abeeaf895811e26cdde032cf29272d7/configure_sharp_stop.log) | :heavy_check_mark: PASS
[run_nccl_test.log](https://gist.githubusercontent.com/mellanox-github/f88dbb2daac122e356460f7e9edc8f85/raw/f37a701cb757d4c7a4526fec8f53f1ad0fe33823/run_nccl_test.log) | :heavy_check_mark: PASS

**Comment by mellanox-github** [comment]
Mellanox CI PASSed.

**Comment by mellanox-github** [comment]
Mellanox CI FAILed.

**Comment by Sergei-Lebedev** [comment]
bot:mlx:test

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI status: <b>PASS</b#  > (click for details)</summary>

**Note:** the artefacts will be deleted after 17-Feb-2020

Artefact file | Status
-- | --
[build_nccl_rdma_sharp_plugins.log](https://gist.githubusercontent.com/mellanox-github/3a265b126730af4be411540a8df7f74f/raw/cf9e2530e2b09984ce7ce9fca38deb5632222d88/build_nccl_rdma_sharp_plugins.log) | :heavy_check_mark: PASS
[configure_sharp_startup.log](https://gist.githubusercontent.com/mellanox-github/3a265b126730af4be411540a8df7f74f/raw/22294bc7388af227c6252e90bd8e2a1af76ad0b7/configure_sharp_startup.log) | :heavy_check_mark: PASS
[configure_sharp_stop.log](https://gist.githubusercontent.com/mellanox-github/3a265b126730af4be411540a8df7f74f/raw/d41511f391419460a57f00000897dc689b861299/configure_sharp_stop.log) | :heavy_check_mark: PASS
[run_nccl_test.log](https://gist.githubusercontent.com/mellanox-github/3a265b126730af4be411540a8df7f74f/raw/2c6b665ad79da2d0bf34da9ecdb80f5b5435f78b/run_nccl_test.log) | :heavy_check_mark: PASS

**Comment by mellanox-github** [comment]
Mellanox CI PASSed.

File: src/Makefile.am [hunk 3]
```diff
@@ -7,9 +7,17 @@ lib_LTLIBRARIES     = libnccl-net.la
...
 libnccl_net_la_CFLAGS   = $(CFLAGS) -I$(top_srcdir)/include -DGNU_SOURCE
 libnccl_net_la_LIBADD   = -lsharp_coll -lcudart_static
-libnccl_net_la_LDFLAGS   = $(LDFLAGS)
+libnccl_net_la_LDFLAGS  = $(LDFLAGS)
 
 libnccl_net_la_SOURCES = \
 	ibvwrap.c \
 	utils.c	\
-	plugin.c
+    ib_plugin.c \
```
Comment (bureddy) [inline src/Makefile.am]: alignment?

File: src/ucx_plugin.c [hunk 4]
```diff
@@ -0,0 +1,872 @@
+
+/*************************************************************************
+ * Copyright (c) 2016-2018, NVIDIA CORPORATION. All rights reserved.
+ * Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+ *
+ * See LICENSE.txt for license information
+ ************************************************************************/
+
+
+
```
Comment (bureddy) [inline src/ucx_plugin.c]: remove 1 empty line

**Comment by bureddy** [comment]
@Sergei-Lebedev  should we move some of the common functions in IB and UCX plugins to some common file?

**Comment by Sergei-Lebedev** [comment]
Was added as PR series #27 #28 #29

**Pull request closed without merging**